src/aem/azext_aem/custom.py

```diff
@@ -0,0 +1,546 @@
+from knack.util import CLIError
+from knack.log import get_logger
+
+from azure.cli.core.profiles import ResourceType, get_sdk
+from azure.cli.core.commands.client_factory import get_mgmt_service_client, get_data_service_client
+
+logger = get_logger(__name__)
+
+aem_extension_info = {
+    'Linux': {
+        'publisher': 'Microsoft.OSTCExtensions',
+        'name': 'AzureEnhancedMonitorForLinux',
+        'version': '3.0'
+    },
+    'Windows': {
+        'publisher': 'Microsoft.AzureCAT.AzureEnhancedMonitoring',
+        'name': 'AzureCATExtensionHandler',
+        'version': '2.2'
+    }
+}
+
+
+def set_aem(cmd, resource_group_name, vm_name, skip_storage_analytics=False):
+    aem = EnhancedMonitoring(cmd, resource_group_name, vm_name,
+                             vm_client=get_mgmt_service_client(cmd.cli_ctx, ResourceType.MGMT_COMPUTE),
+                             storage_client=get_mgmt_service_client(cmd.cli_ctx, ResourceType.MGMT_STORAGE),
+                             skip_storage_analytics=skip_storage_analytics)
+    aem.enable()
+
+
+def delete_aem(cmd, resource_group_name, vm_name):
+    aem = EnhancedMonitoring(cmd, resource_group_name, vm_name,
+                             vm_client=get_mgmt_service_client(cmd.cli_ctx, ResourceType.MGMT_COMPUTE),
+                             storage_client=None)
+    aem.delete()
+
+
+def verify_aem(cmd, resource_group_name, vm_name, wait_time_in_minutes=15, skip_storage_check=False):
+    aem = EnhancedMonitoring(cmd, resource_group_name, vm_name,
+                             vm_client=get_mgmt_service_client(cmd.cli_ctx, ResourceType.MGMT_COMPUTE),
+                             storage_client=get_mgmt_service_client(cmd.cli_ctx, ResourceType.MGMT_STORAGE))
+    aem.verify(skip_storage_check, wait_time_in_minutes)
+
+
+class EnhancedMonitoring(object):
+    def __init__(self, cmd, resource_group, vm_name, vm_client,
+                 storage_client, skip_storage_analytics=None):
+        self._vm_client = vm_client
+        self._storage_client = storage_client
+        self._resource_group = resource_group
+        self._cmd = cmd
+        self._vm = vm_client.virtual_machines.get(resource_group, vm_name, expand='instanceView')
+        # TODO: put it under api-version check on 2016-04-30preview+
+        self._extension = (aem_extension_info['Linux'] if bool(self._vm.os_profile.linux_configuration)
+                           else aem_extension_info['Windows'])
+        self._skip_storage_analytics = skip_storage_analytics
+
+    def enable(self):
+        pub_cfg, pri_cfg = self._build_extension_cfgs(self._get_disk_info())
+        VirtualMachineExtension = self._cmd.get_models('VirtualMachineExtension')
+        existing_ext = self._get_aem_extension()
+        if existing_ext:
+            extension_instance_name = existing_ext.name
+        else:
+            extension_instance_name = self._extension['name']
```

Comment:
nit could be: `extension_instance_name = existing_ext.name if existing_ext else self._extension['name']`?